Option to remove formatting before diff

70ray opened the pull request:
new file unformat.inc with function to remove formatting
and either completely unwrap or keep line structure but
remove leading and trailing spaces

diff.php: add option to remove formatting (not unwrap)
before compare.
If at least one round is F then show link option to remove
formatting or not.
Default is remove for P to F, keep for F to F

new file page_compare.php to compare pages with format
removed and unwrapped between selected rounds for a
project, either all pages or for the users pages in either
the first or second round. Makes a list of page image names
which link to diff.php

project.php: link added to page_compare.php if project has
reached formatting stage

cpeel commented on pinc/unformat.inc:
```diff
@@ -0,0 +1,42 @@
+<?php
+function is_format_round($label)
```
We do this "is this round a formatting/proofreading round" in several places (eg: round_project_listing_presort() in `pinc/stages.inc`). We should move this into the Round class in `pinc/RoundDescriptor.inc`, something like:
```
    function is_proofreading_round()
    {
        return startswith($this->id, 'P');
    }

    function is_formatting_round()
    {
        return startswith($this->id, 'F');
    }
```

jmdyck replied:
`RoundDescriptor.inc` is generic, but the P/F distinction is pgdp-specific. Putting the code in `stages.inc` would be more appropriate, because that's completely pgdp-specific.

cpeel commented on pinc/unformat.inc:
```diff
@@ -0,0 +1,42 @@
+    return "F" == $label[0];
+}
+
+function remove_formatting($text, $unwrap)
```
This probably best belongs in `pinc/DPage.inc` since it manipulates the data for a page.

cpeel commented on project.php:
```diff
@@ -694,7 +695,14 @@ function do_project_info_table()
-                $detail = "<a href='$url'>$blurb</a> &gt;&gt;<a href='$url2'>$blurb2</a>&lt;&lt;";
+                $detail = "<a href='$url'>$blurb</a> | <a href='$url2'>$blurb2</a>";
+                $state_index = array_flip($PROJECT_STATES_IN_ORDER);
+                if($state_index[$project->state] >= $state_index[PROJ_F1_AVAILABLE])
```
Using PROJ_F1_AVAILABLE is non-ideal as in theory different sites can customize the different rounds. It would be more round-about, but probably overall better, if you did something like this (untested) within the Project class:
```
    public function has_been_through_formatting_round()
    {
        global $PROJECT_STATES_IN_ORDER;
        $state_index = array_search($this->state, $PROJECT_STATES_IN_ORDER);
        for($i=$state_index; $i>=0; $i--)
        {
            $round = get_Round_for_project_state($PROJECT_STATES_IN_ORDER[$i]);
            if($round)
            {
                if($round->is_formatting_round())
                    return TRUE;
            }
        }
        return FALSE;
    }
```

cpeel commented on tools/project_manager/diff.php:
```diff
@@ -61,6 +62,19 @@
 
+$L_format = is_format_round($L_label);
+$R_format = is_format_round($R_label);
+$format = array_get($_GET, "format", "default");
```
Move this to the top of the file with the other $_GET validation. It should also be using `get_enumerated_param()` with a default parameter instead of a blanket array_get().

cpeel commented on tools/project_manager/diff.php:
```diff
@@ -61,6 +62,19 @@
+$L_format = is_format_round($L_label);
+$R_format = is_format_round($R_label);
+$format = array_get($_GET, "format", "default");
+if($format == "keep")
```
Please use brackets in all if/then/else statements if you're going to use them in any of them.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,182 @@
+
+$projectid = validate_projectID('project', @$_GET['project']);
+
+$project = new Project($projectid );
```
remove space before closing paren.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,182 @@
+    exit();
+}
+
+$L_round_id = array_get($_GET, "L_round_id", "P3");
```
Move all $_GET accesses to the top of the file rather than have them scattered throughout. This allows one place to do user-input validation.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,182 @@
+
+if($state_index[$project->state] < $state_index["{$R_round_id}.proj_avail"])
+{
+    echo "$R_round_id has not started";
```
This needs to use `_()` (and sprintf()).

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,182 @@
+    $condition .= " AND state='$R_round->page_save_state'";
+
+$query = "
+SELECT image, $L_text_column_name, $R_text_column_name
```
Please indent these 4 spaces so it stands out as SQL. Also, move the closing quote to a line by itself. See Project:get_hold_status() in `pinc/Project.inc` for an example.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,182 @@
+        echo "<form action='page_compare.php' method='GET'>
+            <input type='hidden' name='project' value='$projectid'>\n
+            <input type='hidden' name='compare'>\n";
+        echo "<p>Compare ";
```
This can' be translated as-is.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,182 @@
+        $this->echo_page_select('all', _("All pages"));
+        $this->echo_page_select('left', _("My first"));
+        $this->echo_page_select('right', _("My second"));
+        echo "<input type='submit' value='Go'></p></form>\n";
```
"Go" is not translateable.

70ray commented:
Thanks for the comments. I've made some changes as suggested and a few other changes. I made a separate commit, re-ordered to beginning, for the functions to find if it is a proofing or formatting round since it is self-contained and rebased everything else. Not sure if I should edit the original comment to reflect current reality.

70ray commented:
I seem to have closed this request? Re-opening it with new comment:
Option to remove formatting before diff:
New function unformat() in DPage.inc to remove formatting and either completely unwrap or keep line structure but remove leading and trailing spaces.
New function has_entered_formatting_round() in Project class in project.inc.
Diff.php: add option to remove formatting (not unwrap)  before compare. If at least one round is F then show link option to remove  formatting or not. Default is remove for P to F, keep for F to F.
New file page_compare.php to compare pages with format removed and unwrapped between selected rounds for a project, either all pages or for the users pages in either the first or second round. Makes a list of page image names which link to diff.php.
Project.php: link added to page_compare.php if project has reached formatting stage.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,203 @@
+
+function not_started($round_id)
+{
+    global $state_index, $project;
```
It would be better to pass these into the function instead of making them global.

70ray replied:
Done this, I did wonder originally whether to do that but since they are declared at the top level they are not any more global than they would be anyway. I suppose the best way to do it would be to make them members of a class but that would involve changing some of the original code rather than just adding stuff.

70ray replied:
Oops. It is my new page! I'll look at improving this.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,203 @@
+}
+echo "</p>";
+
+function not_started($round_id)
```
This function could use a better name -- particularly since it's doing two things (returning true/false if it hasn't started and also printing something out if it hasn't). Also better to make it a positive-named function instead of a negative. Perhaps `has_project_started_round()`.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,203 @@
+    function radio_string($value, $label)
+    {
+        $checked = ($this->page_set === $value) ? " checked" : "";
+        return "<input type='radio' name='page_set' value=$value$checked>$label\n";
```
$value needs single-quotes around it and $label needs `html_safe()`

cpeel commented on pinc/Project.inc:
```diff
@@ -566,6 +566,23 @@ XML;
+        for($i=$state_index; $i>=0; $i--)
+        {
+            $round = get_Round_for_project_state($PROJECT_STATES_IN_ORDER[$i]);
+            if($round)
```
You could simplify this somewhat by:
```
if($round && is_formatting_round($round))
```
Totally optional.

cpeel replied:
Although you need to include_once `stages.inc` at the top of this file if you're using a function from it.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,203 @@
+
+output_header("$sub_title: $title", NO_STATSBAR);
+
+echo "<h1>$title</h1>\n";
```
I know other pages put the name of the work in the `<h1>` but it's really not helpful. This page isn't focused on the project, it's focused on comparing pages. I would swap these and put "Compare pages with formatting removed" as the `<h1>` and the nameofwork as the `<h2>`.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,203 @@
+echo "<h2>$sub_title</h2>\n";
+
+$project_url = "$code_url/project.php?id=$projectid&amp;expected_state=$state";
+$label = _("Return to Project Page");
```
Remove the newline between this and the echo below, which is the only thing in this file that uses this label, and add one above this line.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,203 @@
+
+if (!$project->pages_table_exists)
+{
+    echo "<p>";
```
Should these be styled with `class='warning'` or `error`? I could go either way on that.

70ray replied:
I copied this block of code from "tools/project_manager/page_detail.php" so should we change it there also or perhaps better to make it a function somewhere?

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,203 @@
+
+// make an array of imagenames of pages with diffs
+$diff_pages = array();
+for($rownum=0; $rownum < $num_rows; $rownum++)
```
Since you don't use $rownum anywhere, this would be better as:
```php
while($page_res = mysqli_fetch_assoc($res))
```

70ray commented:
Thanks Casey, I've made the changes. Sorry about rebasing. If you like I'll do something about the warning for pages_table_exists in page_detail.php.

70ray commented:
I just realised I can change page_compare without worrying about breaking any existing code so I'll look at improving the parameters.

70ray commented:
I've re-structured page_compare to use a class. Unfortunately this will probably make re-reviewing difficult but I think it is an improvement. Also improved error handling for project errors which could output messages before the header and give uncaught exceptions.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,213 @@
+
+    function get_data()
+    {
+        $this->projectid = array_get($_GET, 'project', null);
```
I prefer you still use `validate_projectID('project', @$_GET['project']);` so it matches the pattern elsewhere in the code. Yes this has the problem that we output an error message before we start the page, but I'd prefer we be consistent as much as possible. In the future perhaps we'll change how we do this, but this should never happen unless someone is mucking with URLs in which case it's their fault they are getting invalid HTML returned from the server.

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,213 @@
+        {
+            try
+            {
+                $this->project = new Project($this->projectid);
```
You don't need the try/catch. On the TEST server when we throw exceptions we print the full exception to help debugging. On PROD we catch the exception in a global exception handler and just print the error. See the top of `pinc/base.inc` for the code that does this.

cpeel commented on project.php:
```diff
@@ -694,7 +694,13 @@ function do_project_info_table()
                 $url2 = "$url&amp;select_by_user";
                 $blurb2 = _("Just my pages");
-                $detail = "<a href='$url'>$blurb</a> &gt;&gt;<a href='$url2'>$blurb2</a>&lt;&lt;";
+                $detail = "<a href='$url'>$blurb</a> | <a href='$url2'>$blurb2</a>";
```
We may want to rethink how we're outputting these strings. $blurb1 and $blurb2 point to the same page with different parameters. Presumably we used `>> <<` to draw attention to what most(?) people will want to use -- which is just their pages. Removing that and placing them all in `|`s makes them appear as though they are all equivalent. Perhaps something like: `$blurb1 &middot; <b>$blurb2</b> | $blurb3`? That somewhat mirrors what we do in the navbar.

70ray commented:
Reverted the last commit which changes error handling.
Modified the links in project.php.
Revised the check_pages_table_exists() function and re-used it also in project.php

cpeel commented on tools/project_manager/page_compare.php:
```diff
@@ -0,0 +1,190 @@
+            <input type='hidden' name='project' value='$this->projectid'>\n
+            <input type='hidden' name='compare'>\n";
+        // TRANSLATORS: %1$s and %2$s are selectors e.g. P3, %3$s etc. are radio button options
+        echo '<p>', sprintf(_('Compare %1$s to %2$s for %3$s %4$s %5$s'),
```
I suspect this is going to be challenging to translate well. Moreover "My first" and "My second" are really confusing labels since they really mean "use all of the pages that I proofread in the [first|second] round" (and I had to read the code to figure that out). This would probably be better as a small form with labels rather than a sentence. Perhaps something like:
```
Compare rounds:
Round 1   [P3]
Round 2   [F1]
Show      o All pages
          o Pages I worked on in Round 1
          o Pages I worked on in Round 2
```

70ray replied:
It would probably be possible to add some javascript to change the text but seems like adding more complexity.
So it would be like
Compare rounds:   [P3]   [F1]
Show      o All pages
          o Pages I worked on in P3
          o Pages I worked on in F1

cpeel replied:
I don't think you need to do any fancy javascript, I just think you need a form/table with labels you can use to refer to things. Much like we do with the review_work.php page (ie: "work round" and "compare round").

70ray commented:
Made changes as suggested.

cpeel commented on styles/layout.less:
```diff
@@ -881,5 +881,15 @@ table.search-column {
+/* Page compare grid */
+
+.grid-wrapper {
+  display: grid;
```
Indents in this file are 4 spaces, not two.

70ray commented:
Indent in layout.less changed and some rebasing for minor changes and reverted changes.

cpeel commented:
This looks great Ray, thank you for your patience as we worked through my comments (and learned more about github PRs along the way)!

70ray commented:
Thanks for all your help Casey.
